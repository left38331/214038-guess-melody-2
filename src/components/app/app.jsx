import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {ActionCreator} from '../../reducer';
import {WelcomeScreen} from 'components/welcome-screen/welcome-screen';
import {GuessArtist} from 'components/guess-artist/guess-artist';
import {GuessGenre} from 'components/guess-genre/guess-genre';

export class App extends React.PureComponent {
  getScreen() {
    const {mistakes, maxMistakes, questions, onWelcomeScreenClick, step, onUserAnswer} = this.props;

    if (step === -1) {
      return <WelcomeScreen
        errorCount={maxMistakes}
        onStartButtonClick={onWelcomeScreenClick}
      />;
    }

    const currentQuestion = questions[step];

    if (currentQuestion) {
      switch (currentQuestion.type) {
        case `genre`:
          return <GuessGenre
            mistakes={mistakes}
            question={currentQuestion}
            onUserAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, maxMistakes)}
          />;

        case `artist`:
          return <GuessArtist
            mistakes={mistakes}
            question={currentQuestion}
            onUserAnswer={(userAnswer) => onUserAnswer(userAnswer, currentQuestion, mistakes, maxMistakes)}
          />;
      }
    }

    return null;
  }

  render() {
    return this.getScreen();
  }
}

App.propTypes = {
  maxMistakes: PropTypes.number.isRequired,
  mistakes: PropTypes.number.isRequired,
  step: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(PropTypes.object),
  onWelcomeScreenClick: PropTypes.func.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  step: state.step,
  mistakes: state.mistakes,
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incrementStep()),
  onUserAnswer: (userAnswer, question, mistakes, maxMistakes) => {
    dispatch(ActionCreator.incrementStep());
    dispatch(ActionCreator.incrementMistake(
        userAnswer,
        question,
        mistakes,
        maxMistakes
    ));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
