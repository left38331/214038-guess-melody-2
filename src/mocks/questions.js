const questions = [
  {
    type: `artist`,
    song: {
      artist: `John Snow`,
      src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
    },
    answers: [
      {
        id: 1,
        picture: `http://placehold.it/134x134`,
        artist: `John Snow`,
      },
      {
        id: 2,
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        id: 3,
        picture: `http://placehold.it/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        id: 1,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
        genre: `rock`,
      },
      {
        id: 2,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/8/89/My_Kind_Of_Town_clip.ogg/My_Kind_Of_Town_clip.ogg.mp3`,
        genre: `pop`,
      },
      {
        id: 3,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/4/40/Madonna_-_You%27ll_See.ogg/Madonna_-_You%27ll_See.ogg.mp3`,
        genre: `jazz`,
      },
      {
        id: 4,
        src: `https://upload.wikimedia.org/wikipedia/ru/transcoded/d/d1/Rammstein_-_Keine_Lust.ogg/Rammstein_-_Keine_Lust.ogg.mp3`,
        genre: `rock`,
      },
    ],
  },
  {
    type: `artist`,
    song: {
      artist: `John Snow`,
      src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
    },
    answers: [
      {
        id: 1,
        picture: `http://placehold.it/134x134`,
        artist: `John Snow`,
      },
      {
        id: 2,
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        id: 3,
        picture: `http://placehold.it/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        id: 1,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
        genre: `rock`,
      },
      {
        id: 2,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/8/89/My_Kind_Of_Town_clip.ogg/My_Kind_Of_Town_clip.ogg.mp3`,
        genre: `pop`,
      },
      {
        id: 3,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/4/40/Madonna_-_You%27ll_See.ogg/Madonna_-_You%27ll_See.ogg.mp3`,
        genre: `jazz`,
      },
      {
        id: 4,
        src: `https://upload.wikimedia.org/wikipedia/ru/transcoded/d/d1/Rammstein_-_Keine_Lust.ogg/Rammstein_-_Keine_Lust.ogg.mp3`,
        genre: `rock`,
      },
    ],
  },
  {
    type: `artist`,
    song: {
      artist: `John Snow`,
      src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
    },
    answers: [
      {
        id: 1,
        picture: `http://placehold.it/134x134`,
        artist: `John Snow`,
      },
      {
        id: 2,
        picture: `http://placehold.it/134x134`,
        artist: `Jack Daniels`,
      },
      {
        id: 3,
        picture: `http://placehold.it/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        id: 1,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3`,
        genre: `rock`,
      },
      {
        id: 2,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/8/89/My_Kind_Of_Town_clip.ogg/My_Kind_Of_Town_clip.ogg.mp3`,
        genre: `pop`,
      },
      {
        id: 3,
        src: `https://upload.wikimedia.org/wikipedia/en/transcoded/4/40/Madonna_-_You%27ll_See.ogg/Madonna_-_You%27ll_See.ogg.mp3`,
        genre: `jazz`,
      },
      {
        id: 4,
        src: `https://upload.wikimedia.org/wikipedia/ru/transcoded/d/d1/Rammstein_-_Keine_Lust.ogg/Rammstein_-_Keine_Lust.ogg.mp3`,
        genre: `rock`,
      },
    ],
  }
];

const settings = {
  gameTime: 300,
  errorCount: 3
};

export {questions, settings};
