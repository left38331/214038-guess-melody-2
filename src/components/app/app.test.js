import React from 'react';
import renderer from 'react-test-renderer';
import {createStore} from 'redux';
import {Provider} from 'react-redux';

import {App} from './app';
import {questions} from '../../mocks/questions';

it(`render correctly first screen`, () => {
  const props = {
    mistakes: 0,
    maxMistakes: 3,
    step: -1,
    onWelcomeScreenClick: () => {},
    onUserAnswer: () => {},
    isAuthorizationRequired: true,
    requireAuthorization: () => {},
    onSubmit: () => {},
  };

  const mainComponent = renderer.create(<App {...props} />).toJSON();

  expect(mainComponent).toMatchSnapshot();
});

const gameState = {
  step: -1,
  time: 300,
};

const userState = {
  mistakes: 0,
};

const store = createStore(() => ({
  stateUser: userState,
  stateGame: gameState
}));

it(`render correctly second screen`, () => {
  const props = {
    mistakes: 0,
    maxMistakes: 3,
    step: 0,
    onWelcomeScreenClick: () => {},
    onUserAnswer: () => {},
    questions,
    isAuthorizationRequired: true,
    requireAuthorization: () => {},
    onSubmit: () => {},
  };
  const mainComponent = renderer.create(<Provider store={store}>
    <App {...props} />
  </Provider>).toJSON();

  expect(mainComponent).toMatchSnapshot();
});

it(`render correctly third screen`, () => {
  const props = {
    mistakes: 0,
    maxMistakes: 3,
    step: 1,
    onWelcomeScreenClick: () => {},
    onUserAnswer: () => {},
    questions,
    isAuthorizationRequired: true,
    requireAuthorization: () => {},
    onSubmit: () => {},
  };
  const mainComponent = renderer.create(<Provider store={store}>
    <App {...props} />
  </Provider>).toJSON();

  expect(mainComponent).toMatchSnapshot();
});
