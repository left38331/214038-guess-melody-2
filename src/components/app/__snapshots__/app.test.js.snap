// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly first screen 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[Function]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      За 5 минут нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      3
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;

exports[`render correctly second screen 1`] = `
<section
  className="game game--artist"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Кто исполняет эту песню?
    </h2>
    <div
      className="game__track"
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3"
          />
        </div>
      </div>
    </div>
    <form
      className="game__artist"
    >
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={1}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-1"
        />
        <label
          className="artist__name"
          htmlFor={1}
        >
          <img
            alt="John Snow"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          John Snow
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={2}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-2"
        />
        <label
          className="artist__name"
          htmlFor={2}
        >
          <img
            alt="Jack Daniels"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          Jack Daniels
        </label>
      </div>
      <div
        className="artist"
      >
        <input
          className="artist__input visually-hidden"
          id={3}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="answer-3"
        />
        <label
          className="artist__name"
          htmlFor={3}
        >
          <img
            alt="Jim Beam"
            className="artist__picture"
            src="http://placehold.it/134x134"
          />
          Jim Beam
        </label>
      </div>
    </form>
  </section>
</section>
`;

exports[`render correctly thrid screen 1`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    />
    ;
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите инди-рок треки
    </h2>
    <form
      className="game__tracks"
      onSubmit={[Function]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="https://upload.wikimedia.org/wikipedia/en/transcoded/7/7d/ACDC_Itsalongway.ogg/ACDC_Itsalongway.ogg.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id={1}
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-1"
          />
          <label
            className="game__check"
            htmlFor={1}
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="https://upload.wikimedia.org/wikipedia/en/transcoded/8/89/My_Kind_Of_Town_clip.ogg/My_Kind_Of_Town_clip.ogg.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id={2}
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-2"
          />
          <label
            className="game__check"
            htmlFor={2}
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="https://upload.wikimedia.org/wikipedia/en/transcoded/4/40/Madonna_-_You%27ll_See.ogg/Madonna_-_You%27ll_See.ogg.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id={3}
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-3"
          />
          <label
            className="game__check"
            htmlFor={3}
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="https://upload.wikimedia.org/wikipedia/ru/transcoded/d/d1/Rammstein_-_Keine_Lust.ogg/Rammstein_-_Keine_Lust.ogg.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            checked={false}
            className="game__input visually-hidden"
            id={4}
            name="answer"
            onChange={[Function]}
            type="checkbox"
            value="answer-4"
          />
          <label
            className="game__check"
            htmlFor={4}
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;
